### DISCUSSION TITLE
Code workaround to fix this compile error? "error C3487: 'pybind11::array_t<double64_t,16>': all return expressions must deduce to the same type: previously it was 'pybind11::array_t<float,16>'"

### DISCUSSION THREAD

**Discussion by sharpe5** [pybind/pybind11#3993] [Q&A]
In [NumPy](https://numpy.org/), there is the function with this Python signature:

```
x1 = np.ones(100, np.float32)   # Returns 100 ones as float32.
x2 = np.ones(100, np.double64)  # Returns 100 ones as double64.
```

How can I implement this in C++ in PyBind11? It seems easy, but when I code it up, it gives the error in the subject.

Here is the code snippet. The code compiles and works perfectly if I comment out the `case 'd'`, or I comment out the `case 'f'`. However, it won't work if I include both.

```
m.def("zeros", [](int64_t shape, py::object dtype)
		{
			py::dtype x;
			py::dtype y = x.from_args(dtype);
			char ntype = y.char_(); // For example, `float` is 'f', `double` 'd', `int` 'i', and `long` is 'l'.
			switch (ntype) {
				case 'f':
					return toPyArray<float32_t>(std::move(ones<float32_t>(shape, njobs)));
				case 'd':
					return toPyArray<double64_t>(std::move(ones<double64_t>(shape, njobs)));
				default:
					throw new std::exception("Error E49840. Unknown datatype.");
			}					
		},
		"Returns vector of ones.",
			py::arg("shape"),
			py::arg("dtype"),
			py::arg("njobs"));
```

I am wondering if pybind11 is flexible enough to do this? Or would it require complicated internal changes to do so?

What I have tried so far:
* A generic signature of `py::object` and `py:buffer_info`. However, PyBind11 gives an error that it cannot convert the return type.

**Comment by designerzim** [comment]
Similar issue where either a pointer or NULL is returned by the C++ function, but the binding can't compile because NULL is seen as an int instead of a typed pointer.
